### PR TITLE
Rewrite constexpr with template metaprogram

### PR DESCRIPTION
**Pull request by kkoopa**
Get rid of constexpr

### CODE REVIEW

**Comment by kkoopa** [comment]
I have not tested it on VS2013. I do not have it installed, but I do not see why it would not work there.

**Comment by kfarnung** [comment]
Ok, let's not worry about it then. IIRC node dropped support for building extensions using VS2013 anyway.

**Comment by mhdawson** [comment]
CI run on master: https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api/

hmm seems to fail I think it needs to be rebased to be able to validate. @kfarnung do you think this is still something we should pull in (it has been a long time).

**Comment by kfarnung** [comment]
I guess my question is what benefit we get from switching away from constexpr.  I'm not a compiler expert, but presumably there's some compatibility or other benefit from this change.  @kkoopa can you elaborate here?

I'm not opposed to this change assuming we can get the code rebased and passing the CI.

**Comment by kkoopa** [comment]
IIRC, it was for some compatibility reason. I think it pertained to older versions of Visual Studio not fully supporting all constexpr constructs.

**Comment by DaAitch** [comment]
> I guess my question is what benefit we get from switching away from constexpr. I'm not a compiler expert, but presumably there's some compatibility or other benefit from this change. @kkoopa can you elaborate here?
> 
> I'm not opposed to this change assuming we can get the code rebased and passing the CI.

I think it's a win-win situation:
1. the biggest benefit is that for a wrong T it simply does not compile
2. old compilers (before C++11) may not support constexpr, so a check is needed.

Old implementation will always compile and evaluate at runtime:
- using `constexpr`: may inline at compile time that `TypedArrayTypeForPrimitiveType<T>()` is substituted with a e.g. value of `napi_uint16_array`
- not using `constexpr`: may also inline at compile time, since an expression like `return false ? napi_int8_array : false ? napi_uint8_array : true ? napi_int16_array ...` will be optimized

thumbs up for resolving merge conflicts :smile:

**Comment by kkoopa** [comment]
I looked at the code and can clarify my previous statement: It was for getting rid of the check for `NAPI_HAS_CONSTEXPR` which could lead to runtime evaluation instead of compile-time evaluation if not set, which was related to some version of Visual Studio that did not support `constexpr` properly.

**Comment by DaAitch** [comment]
@kkoopa I think you only have to resolve the conflicts so that it's reviewed again :+1:

**Comment by kkoopa** [comment]
> Isn't this pretty much overlapping with #128 ?
I do not remember, but they are probably not overlapping, as I would not have made #128 otherwise.

File: napi-inl.h [hunk 1]
```diff
@@ -1548,6 +1548,38 @@ inline Napi::ArrayBuffer TypedArray::ArrayBuffer() const {
   return Napi::ArrayBuffer(_env, arrayBuffer);
 }
 
+#define X(TYPE)                                                                \
+  template <> struct TypedArray::type_translator<TYPE##_t> {                   \
+    static const napi_typedarray_type value = napi_##TYPE##_array;             \
+  }
+
+X(int8);
```
[the remark this answers is not in the document]
Comment (kkoopa) [reply napi-inl.h]: Yes, that would be better since there are several special cases now.

**Pull request closed without merging**